index.js

```diff
@@ -137,6 +141,33 @@ WPCOM.prototype.sendRequest = function (params, query, body, fn) {
   return sendRequest.call(this, params, query, body, fn)
 };
 
+WPCOM.prototype.wpPromise = function ( callback, params, timeout ) {
+  var timeout = timeout || DEFAULT_ASYNC_TIMEOUT,
+      timer;
+
+  var action = new Promise( function ( resolve, reject ) {
+    var curriedCallback = params ? function(fn) { callback(params, fn); } : function(fn) { callback(fn); };
+
+    curriedCallback( function ( error, data ) {
+      clearTimeout( timer );
+
+      if ( error ) {
+        reject( error );
+      } else {
+        resolve( data );
+      }
+    } );
+  } );
+
+  var timeout = new Promise( function( resolve, reject ) {
+    timer = setTimeout( function() {
+      reject( new Error( 'Action timed out while waiting for return.' ) );
```

Comment:
@retrofox: I'm not sure what is best here - a `TimeoutError`, if such a thing exists, would be great so we can distinguish between a timeout and a rejection from the action Promise.